### PR TITLE
Custom aspect ratio for OrthoCamera

### PR DESCRIPTION
**Pull request by fuchsnj**

### CODE REVIEW

File: amethyst_utils/src/ortho_camera.rs [hunk 1]
```diff
@@ -5,39 +5,93 @@ use amethyst_core::specs::{
...
-/// to preferences in the "mode" field.
-/// It tries as much as possible to adjust the camera so that the world's coordinate (0, 0) is at the bottom left and
-/// (1, 1) is at the top right of the window.
-/// You must add the `CameraNormalOrthoSystem` to your dispatcher for this to take effect.
+/// to preferences in the "mode" and "world_coordinates" fields.
+/// It adjusts the camera so that the camera's world coordinates are always visible.
+/// You must add the `CameraNormalOrthoSystem` to your dispatcher for this to take effect (no dependencies required).
 #[derive(Default, Debug, Clone, Serialize, Deserialize)]
-pub struct CameraNormalOrtho {
+pub struct OrthoCamera {
```
Comment (jojolepro) [inline amethyst_utils/src/ortho_camera.rs]: Please keep Camera at the start of the name. When the feature was first merged in, the reviewers choosed to have Camera at the start of the name to help autocompletion.

File: docs/CHANGELOG.md [hunk 2]
```diff
@@ -28,6 +28,7 @@ The format is based on [Keep a Changelog][kc], and this project adheres to
 * New vertices() method on MeshCreator trait. ([#946])
 * Support for text alignment (align left, center, right). ([#965])
 * Support for multiline text. ([#965])
+* Added custom aspect ratio to OrthoCamera. ([#983])
```
Comment (jojolepro) [inline docs/CHANGELOG.md]: You need to also link the pull request like so:
[#983]: https://github.com/amethyst/amethyst/pull/983/

File: amethyst_utils/src/ortho_camera.rs (hunk 1 above)
Comment (fuchsnj) [reply amethyst_utils/src/ortho_camera.rs]: I don't really agree with that, but I changed it.

File: docs/CHANGELOG.md (hunk 2 above)
Comment (jojolepro) [reply docs/CHANGELOG.md]: :+1:

**Comment by Rhuagh** [comment]
Please squash the commits.

**Comment by LucioFranco** [comment]
@fuchsnj we use @bors[bot] to do the merging and it doesn't squash it does a straight merge. So when PR's get a bit long or have commits about fixes we usually want to squash them. 😄

**Comment by fuchsnj** [comment]
Travis failed because it got some 503's downloading dependencies. What's the standard way of restarting the build? I'm just going to push an empty commit then squash again... is there a bors command or something that can be used?

**Comment by jojolepro** [comment]
You need to push a commit for it to restart

File: amethyst_utils/src/ortho_camera.rs [hunk 3]
```diff
@@ -5,39 +5,93 @@ use amethyst_core::specs::{
...
+            bottom: 0.0,
+            top: 1.0,
+        }
+    }
+
+    /// width / height of the desired camera coordinates
+    pub fn aspect_ratio(&self) -> f32 {
+        self.width() / self.height()
+    }
+    /// size of the x-axis
```
Comment (jojolepro) [inline amethyst_utils/src/ortho_camera.rs]: Missing space before the doc comment

File: amethyst_utils/src/ortho_camera.rs [hunk 4]
```diff
@@ -5,39 +5,93 @@ use amethyst_core::specs::{
...
+    pub fn width(&self) -> f32 {
+        self.right - self.left
+    }
+    /// size of the y-axis
```
Comment (jojolepro) [inline amethyst_utils/src/ortho_camera.rs]: Also missing a line before

File: amethyst_utils/src/ortho_camera.rs [hunk 5]
```diff
@@ -5,39 +5,93 @@ use amethyst_core::specs::{
...
+impl CameraOrtho {
+    /// Creates a Camera that maintains window coordinates of (0,0) in the bottom left, and (1,1) at the top right
+    pub fn normalized(mode: CameraNormalizeMode) -> CameraOrtho {
+        CameraOrtho {
+            mode,
+            world_coordinates: Default::default(),
+        }
+    }
+
+    pub fn camera_offsets(&self, window_aspect_ratio: f32) -> (f32, f32, f32, f32) {
```
Comment (jojolepro) [inline amethyst_utils/src/ortho_camera.rs]: Missing doc on pub function

**Comment by jojolepro** [comment]
Hello, can you fix my latest review? This is a cool feature I would like to get into the engine :)

Thanks!

**Comment by torkleyy** [comment]
Seems to be fixed?

**Comment by torkleyy** [comment]
Oh, just not rebased.

File: amethyst_utils/src/ortho_camera.rs [hunk 6]
```diff
@@ -5,39 +5,96 @@ use amethyst_core::specs::{
 use amethyst_core::Axis2;
 use amethyst_renderer::{Camera, ScreenDimensions};
 
+/// The coordinates that `CameraOrtho` will keep visible in the window
+#[derive(Debug, Clone, Serialize, Deserialize)]
```
Comment (azriel91) [inline amethyst_utils/src/ortho_camera.rs]: Could you add `Copy` and `PartialEq` as well?
That'll make it more ergonomic to use, and easier to assert in tests

File: amethyst_utils/src/ortho_camera.rs [hunk 7]
```diff
@@ -5,39 +5,96 @@ use amethyst_core::specs::{
...
+    pub fn normalized() -> CameraOrthoWorldCoordinates {
+        CameraOrthoWorldCoordinates {
+            left: 0.0,
+            right: 1.0,
+            bottom: 0.0,
+            top: 1.0,
+        }
+    }
+
+    /// width / height of the desired camera coordinates
```
Comment (azriel91) [inline amethyst_utils/src/ortho_camera.rs]: pedantic

```rust
/// Returns the `width / height` ratio of the desired camera coordinates.
```

File: amethyst_utils/src/ortho_camera.rs [hunk 8]
```diff
@@ -5,39 +5,96 @@ use amethyst_core::specs::{
...
+    pub fn aspect_ratio(&self) -> f32 {
+        self.width() / self.height()
+    }
+
+    /// size of the x-axis
```
Comment (azriel91) [inline amethyst_utils/src/ortho_camera.rs]:
```rust
/// Returns the size of the X axis.
```

File: amethyst_utils/src/ortho_camera.rs [hunk 9]
```diff
@@ -5,39 +5,96 @@ use amethyst_core::specs::{
...
+    pub fn width(&self) -> f32 {
+        self.right - self.left
+    }
+
+    /// size of the y-axis
```
Comment (azriel91) [inline amethyst_utils/src/ortho_camera.rs]:
```rust
/// Returns the size of the X axis.
```

also, we should document that `top` should be more positive than `bottom` in the struct definition, and perhaps add a `new(..)` function that validates that

File: amethyst_utils/src/ortho_camera.rs [hunk 10]
```diff
@@ -5,39 +5,96 @@ use amethyst_core::specs::{
...
+
 /// `Component` attached to the camera's entity that allows automatically adjusting the camera's matrix according
-/// to preferences in the "mode" field.
-/// It tries as much as possible to adjust the camera so that the world's coordinate (0, 0) is at the bottom left and
-/// (1, 1) is at the top right of the window.
-/// You must add the `CameraNormalOrthoSystem` to your dispatcher for this to take effect.
+/// to preferences in the "mode" and "world_coordinates" fields.
+/// It adjusts the camera so that the camera's world coordinates are always visible.
+/// You must add the `CameraNormalOrthoSystem` to your dispatcher for this to take effect (no dependencies required).
 #[derive(Default, Debug, Clone, Serialize, Deserialize)]
```
Comment (azriel91) [inline amethyst_utils/src/ortho_camera.rs]: might as well add `PartialEq` here 😄

File: amethyst_utils/src/ortho_camera.rs [hunk 11]
```diff
@@ -5,39 +5,96 @@ use amethyst_core::specs::{
...
     type Storage = DenseVecStorage<Self>;
 }
 
 /// Settings that decide how to scale the camera's matrix when the aspect ratio changes.
 #[derive(Debug, Copy, Clone, Serialize, Deserialize)]
 pub enum CameraNormalizeMode {
-    /// Using an aspect ratio of 1:1, tries to adjust the matrix values of the camera so
-    /// that the direction opposite to the stretch_direction always have a world size of 1.
+    /// Using the aspect ratio from the world coordinates for this camera, tries to adjust the matrix values of the
+    /// camera so that the direction opposite to the stretch_direction always have a world size of 1.
```
Comment (azriel91) [inline amethyst_utils/src/ortho_camera.rs]: can you replace "opposite" with "orthogonal"? I thinks that's more accurate

**Comment by torkleyy** [comment]
@fuchsnj In case you don't have time to fix this until Friday, would you like me to finish it for you?

**Comment by jojolepro** [comment]
I fixed the stuff and things.

bors r+

**Comment by jojolepro** [comment]
bors r+

Was missing a PartialEq impl

**Pull request merged**